### ISSUE TITLE
cross-compilation from GNU/Linux/x86_64 to MinGW/i686 broken

### ISSUE THREAD

**Issue by svillemot** [xianyi/OpenBLAS#1674]
Hi,

When the host is GNU/Linux x86_64 and the target is MinGW/i686, then the build system does not detect that this is a cross compilation.

This is because line 230 of `c_check` reads:
``
 $cross = 0 if (($hostarch eq "x86_64") && ($architecture eq "x86"));
``
This test should set `$cross` to 0 only if the previous test on operating system (`$os ne $hostos`) did not already detect cross-compilation.

Best,

**Comment by brada4** [comment]
works for me just fine (2.20, 3.0 , 3.1) and particular code is from original gotoblas....

`make CC=mingw-....-gcc FC=mingw....-gfortran DYNAMIC_ARCH=1 HOSTCC=gcc`

EDIT: yes, produces workingwin32  BLAS DLL as a result.

**Comment by martin-frbg** [comment]
@svillemot what `make' options did you use ?

**Comment by brada4** [comment]
I suspect that option means AS=as and nothing more, x86_64 as generates i686 code just fine

**Comment by svillemot** [comment]
@martin-frbg 
```
DYNAMIC_ARCH=1
CC = i686-w64-mingw32-gcc
FC = i686-w64-mingw32-gfortran
HOSTCC=gcc
USE_THREAD = 1
NUM_THREADS = 2
NO_SHARED = 1
NO_LAPACKE = 1
```

**Comment by brada4** [comment]
Works for me with centos7? Can you get output of failure?

**Comment by svillemot** [comment]
Actually I should have been more specific. Building the library succeeds, but then the build process wants to run the tests:
```
OPENBLAS_NUM_THREADS=1 OMP_NUM_THREADS=1 ./sblat1
run-detectors: unable to find an interpreter for ./sblat1
make[1]: *** [Makefile:8: level1] Error 2
```
which of course fails because I can't run Windows binaries on my GNU/Linux box.

I can also verify that the generated `Makefile.conf` does not contain `CROSS=1`.

As explained in my first message, the reason is simply that `c_check` assumes that compiling from x86_64 to x86 is *never* a cross compilation. While in a sense this is true when the host and the target are the same OS, this is not true when cross-compiling from GNU/Linux to MinGW.

**Comment by brada4** [comment]
Got you. The absolute platform difference should prevail.

**Comment by martin-frbg** [comment]
I wonder what platform(s) we could break if we simply move the "$os ne $hostos" check below the "$architecture ne $hostarch" one..

**Comment by brada4** [comment]
We just fix this particular case? (and mips and arm cousins of it)

Actually OpenBSD amd64 does not have provisions to run 32bit multiarch, that would even invalidate assumption that same system can run lower arch tests.

**Comment by martin-frbg** [comment]
Should be fixed by #1746

**Comment by svillemot** [comment]
Thanks!
